### PR TITLE
Improve Nominatim provider

### PR DESCRIPTION
**Pull request by jbelien**
- Migration from format XML to JSON (v2)
- Add `countrycodes` geocoding query parameter
- Add `viewbox` (and `bounded`) geocoding query parameter(s)

See https://wiki.openstreetmap.org/wiki/Nominatim

I paid attention to not create any breaking change.

### CODE REVIEW

File: src/Provider/Nominatim/Model/NominatimAddress.php [hunk 1]
```diff
@@ -73,24 +73,34 @@ public function withAttribution(string $attribution = null): self
     /**
      * @return null|string
      */
+    public function getCategory()
+    {
+        return $this->category;
+    }
+
     public function getClass()
```
Comment (Nyholm) [inline src/Provider/Nominatim/Model/NominatimAddress.php]: We should mark this function as `@deprecated`

File: src/Provider/Nominatim/Model/NominatimAddress.php [hunk 2]
```diff
@@ -73,24 +73,34 @@ public function withAttribution(string $attribution = null): self
...
+    public function withCategory(string $category = null): self
     {
         $new = clone $this;
-        $new->class = $class;
+        $new->category = $category;
 
         return $new;
     }
 
+    public function withClass(string $category = null): self
```
Comment (Nyholm) [inline src/Provider/Nominatim/Model/NominatimAddress.php]: Why do we need this? Dont we rename `class` to `category`?

File: src/Provider/Nominatim/Nominatim.php [hunk 3]
```diff
@@ -54,7 +54,7 @@
      *
      * @return Nominatim
      */
-    public static function withOpenStreetMapServer(HttpClient $client, string $userAgent, string $referer = '')
+    public static function withOpenStreetMapServer(HttpClient $client, string $userAgent, string $referer = ''): self
```
Comment (Nyholm) [inline src/Provider/Nominatim/Nominatim.php]: I think adding this is a BC break. Or is this class `final`?

Comment (jbelien) [reply src/Provider/Nominatim/Nominatim.php]: I think this class could be declared as `final` !
But I can remove the `self`, it doesn't "add" a lot.

File: src/Provider/Nominatim/Model/NominatimAddress.php (hunk 1 above)
Comment (jbelien) [reply src/Provider/Nominatim/Model/NominatimAddress.php]: Done !

File: src/Provider/Nominatim/Nominatim.php (hunk 3 above)
Comment (jbelien) [reply src/Provider/Nominatim/Nominatim.php]: I just checked : the class is (and was) indeed `final` : https://github.com/geocoder-php/Geocoder/blob/master/src/Provider/Nominatim/Nominatim.php#L33

**Comment by jbelien** [comment]
@Nyholm Can I merge this ? :)

File: src/Provider/Nominatim/Model/NominatimAddress.php (hunk 2 above)
Comment (Nyholm) [reply src/Provider/Nominatim/Model/NominatimAddress.php]: Oh, Yes of course. 

The diff was weird. I thought you added a new "withClass" function

**Comment by jbelien** [comment]
Thanks ! 👍

**Pull request merged**